File: services/worker/worker/secrets.py
from __future__ import annotations

import base64

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def decrypt_secret(value: dict, encoded_key: str) -> str:
    if value.get("algorithm") != "aes-256-gcm" or value.get("version") != 1:
        raise ValueError("Unsupported credential encryption format")
    key = base64.b64decode(encoded_key, validate=True)
    if len(key) != 32:
        raise ValueError("CREDENTIAL_ENCRYPTION_KEY must decode to 32 bytes")
    iv = base64.b64decode(value["iv"])
    ciphertext = base64.b64decode(value["ciphertext"])
    tag = base64.b64decode(value["tag"])
    return AESGCM(key).decrypt(iv, ciphertext + tag, None).decode("utf-8")

File: services/worker-go/worker/secrets.go
package main

import (
	"crypto/aes"
	"crypto/cipher"
	"encoding/base64"
	"fmt"
)

func DecryptSecret(value map[string]interface{}, encodedKey string) (string, error) {
	if stringValue(value["algorithm"]) != "aes-256-gcm" || intValue(value["version"]) != 1 {
		return "", fmt.Errorf("unsupported credential encryption format")
	}
	key, err := base64.StdEncoding.DecodeString(encodedKey)
	if err != nil {
		return "", err
	}
	if len(key) != 32 {
		return "", fmt.Errorf("CREDENTIAL_ENCRYPTION_KEY must decode to 32 bytes")
	}
	iv, err := base64.StdEncoding.DecodeString(stringValue(value["iv"]))
	if err != nil {
		return "", err
	}
	ciphertext, err := base64.StdEncoding.DecodeString(stringValue(value["ciphertext"]))
	if err != nil {
		return "", err
	}
	tag, err := base64.StdEncoding.DecodeString(stringValue(value["tag"]))
	if err != nil {
		return "", err
	}
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}
	aead, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}
	payload := append(ciphertext, tag...)
	plain, err := aead.Open(nil, iv, payload, nil)
	if err != nil {
		return "", err
	}
	return string(plain), nil
}
